### PR TITLE
Start re-implementing JavaFxScheduler

### PR DESCRIPTION
**Pull request by thomasnield**
See #27. I am kind of hacking my way through this, hoping to get some understanding on the way. For now I copied the Swing Scheduler and refactored where it was obvious. However I am unsure what to do with `Timer` and `ActionListener` since they both are awt-based, and I need to use the JavaFX equivalents I imagine.

### CODE REVIEW

**Comment by thomasnield** [comment]
Doing a file compare between the old and new Scheduler, I think I have some ideas. Will implement by end of the week. @akarnokd I may seek your approval when I'm done if that's okay.

**Comment by thomasnield** [comment]
Did my first pass although unit test is failing. I'm not quite sure what to do with the `Timeline` which I figured out is acting as the equivalent to the awt `Timer`.

**Comment by thomasnield** [comment]
Alright, I think I need some help here. The scheduler test is failing and I'm pretty sure I have a loose end somewhere in the `JavaFxScheduler`. 

I don't think I can get the FxCollections interop factories out with the existing `JavaFxScheduler` because I've had scheduling issues with `ObservableList` events.  @akarnokd and @lestard, if you can ever make time can you look at my trampolining `JavaFxScheduler` and see if there is anything obviously wrong?

https://github.com/thomasnield/RxJavaFX/blob/scheduler_fix/src/main/java/rx/schedulers/JavaFxScheduler.java

**Comment by thomasnield** [comment]
Here is @akarnokd's [Swing Scheduler implementation](https://github.com/ReactiveX/RxSwing/pull/24/files) I'm trying to mimic.

**Comment by thomasnield** [comment]
I may go check out Tomas Mikula's work and see what he did with [ReactFX](https://github.com/TomasMikula/ReactFX/blob/931174b43d07d6509b8635784dfefbc6e0e1f0cb/reactfx/src/main/java/org/reactfx/ScheduledExecutorServiceTimer.java)

**Comment by thomasnield** [comment]
Okay, I have a feeling I'm 98% there. I have implemented this verbatim to the Swing scheduler and found all the proper JavaFX equivalents. 

``` java
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.util.Duration;
import rx.Scheduler;
import rx.Subscription;
import rx.functions.Action0;
import rx.subscriptions.BooleanSubscription;
import rx.subscriptions.CompositeSubscription;
import rx.subscriptions.SerialSubscription;
import rx.subscriptions.Subscriptions;

import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.TimeUnit;

/**
 * Executes work on the JavaFx UI thread.
 * This scheduler should only be used with actions that execute quickly.
 */
public final class JavaFxScheduler extends Scheduler {
    private static final JavaFxScheduler INSTANCE = new JavaFxScheduler();

    /* package for unit test */JavaFxScheduler() {
    }

    public static JavaFxScheduler getInstance() {
        return INSTANCE;
    }

    private static void assertThatTheDelayIsValidForTheJavaFxTimer(long delay) {
        if (delay < 0 || delay > Integer.MAX_VALUE) {
            throw new IllegalArgumentException(String.format("The JavaFx timer only accepts non-negative delays up to %d milliseconds.", Integer.MAX_VALUE));
        }
    }

    @Override
    public Worker createWorker() {
        return new InnerJavaFxScheduler();
    }

    private static class InnerJavaFxScheduler extends Worker implements Runnable {

        private final CompositeSubscription tracking = new CompositeSubscription();

        /** Allows cheaper trampolining than invokeLater(). Accessed from EDT only. */
        private final Queue<Runnable> queue = new ConcurrentLinkedQueue<>();
        /** Allows cheaper trampolining than invokeLater(). Accessed from EDT only. */
        private int wip;

        @Override
        public void unsubscribe() {
            tracking.unsubscribe();
        }

        @Override
        public boolean isUnsubscribed() {
            return tracking.isUnsubscribed();
        }

        @Override
        public Subscription schedule(final Action0 action, long delayTime, TimeUnit unit) {
            long delay = Math.max(0,unit.toMillis(delayTime));
            assertThatTheDelayIsValidForTheJavaFxTimer(delay);

            class DualAction implements EventHandler<ActionEvent>, Subscription, Runnable {
                private Timeline timeline;
                final SerialSubscription subs = new SerialSubscription();
                boolean nonDelayed;

                private void setTimer(Timeline timeline) {
                    this.timeline = timeline;
                }

                @Override
                public void handle(ActionEvent event) {
                    run();
                }

                @Override
                public void run() {
                    if (nonDelayed) {
                        try {
                            if (tracking.isUnsubscribed() || isUnsubscribed()) {
                                return;
                            }
                            action.call();
                        } finally {
                            subs.unsubscribe();
                        }
                    } else {
                        timeline.stop();
                        timeline = null;
                        nonDelayed = true;
                        trampoline(this);
                    }
                }

                @Override
                public boolean isUnsubscribed() {
                    return subs.isUnsubscribed();
                }

                @Override
                public void unsubscribe() {
                    subs.unsubscribe();
                }
                public void set(Subscription s) {
                    subs.set(s);
                }
            }

            final DualAction executeOnce = new DualAction();
            tracking.add(executeOnce);

            final Timeline timer = new Timeline(new KeyFrame(Duration.millis(delay), executeOnce));
            executeOnce.setTimer(timer);
            timer.play();

            executeOnce.set(Subscriptions.create(() -> {
                timer.stop();
                tracking.remove(executeOnce);
            }));

            return executeOnce;
        }

        @Override
        public Subscription schedule(final Action0 action) {
            final BooleanSubscription s = BooleanSubscription.create();
            Runnable runnable = () -> {
                try {
                    if (tracking.isUnsubscribed() || s.isUnsubscribed()) {
                        return;
                    }
                    action.call();
                } finally {
                    tracking.remove(s);
                }
            };
            tracking.add(s);

            if (Platform.isFxApplicationThread()) {
                if (trampoline(runnable)) {
                    return Subscriptions.unsubscribed();
                }
                else {
                    queue.offer(runnable);
                    Platform.runLater(this);
                }
            }

            // wrap for returning so it also removes it from the 'innerSubscription'
            return Subscriptions.create(() -> tracking.remove(s));
        }
        /**
         * Uses a fast-path/slow path trampolining and tries to run
         * the given runnable directly.
         * @param runnable
         * @return true if the fast path was taken
         */
        boolean trampoline(Runnable runnable) {
            // fast path: if wip increments from 0 to 1
            if (wip == 0) {
                wip = 1;
                runnable.run();
                // but a recursive schedule happened
                if (--wip > 0) {
                    do {
                        Runnable r = queue.poll();
                        r.run();
                    } while (--wip > 0);
                }
                return true;
            }
            queue.offer(runnable);
            run();
            return false;
        }
        @Override
        public void run() {
            if (wip++ == 0) {
                do {
                    Runnable r = queue.poll();
                    r.run();
                } while (--wip > 0);
            }
        }
    }
}

```

I've studied this in detail and I'm confused why `observeOn()` operators are not working.

``` java
Observable<ActionEvent> bttnEvents =
     JavaFxObservable.fromActionEvents(incrementBttn);

bttnEvents.map(e -> 1).observeOn(Schedulers.io())
    .scan(0,(x, y) -> x + y)
    .observeOn(JavaFxScheduler.getInstance())
    .doOnNext(System.out::println)
```

Oddly I do not get any numbers printed every time I click the button. It's like the events are not being fired.

File: src/main/java/rx/schedulers/JavaFxScheduler.java [hunk 1]
```diff
@@ -46,86 +48,167 @@ public static JavaFxScheduler getInstance() {
...
 
-            });
+            final DualAction executeOnce = new DualAction();
+            tracking.add(executeOnce);
+
+            //final Timeline timer = new Timeline((int) delay, executeOnce);
+            final Timeline timer = new Timeline(new KeyFrame(Duration.millis(delay)));
+
+            executeOnce.setTimer(timer);
+            timer.play();
```
Comment (akarnokd) [inline src/main/java/rx/schedulers/JavaFxScheduler.java:122]: Does the Timeline execute its action on the JavaFX thread? If not, then the `trampoline()`in its current form won't work.

File: src/main/java/rx/schedulers/JavaFxScheduler.java [hunk 2]
```diff
@@ -46,86 +48,167 @@ public static JavaFxScheduler getInstance() {
...
 
-            innerSubscription.add(s);
-            // wrap for returning so it also removes it from the 'innerSubscription'
-            return Subscriptions.create(new Action0() {
-
-                @Override
-                public void call() {
-                    s.unsubscribe();
-                    innerSubscription.remove(s);
+            if (Platform.isFxApplicationThread()) {
```
Comment (akarnokd) [inline src/main/java/rx/schedulers/JavaFxScheduler.java:147]: What if `schedule()` is called from not the platform thread? Looks like you need:

``` java
if (Platform.isFxApplicationThread()) {
    if (trampoline(runnable)) {
        return Subscriptions.unsubscribed();
    }
} else {
    queue.offer(runnable);
    Platform.runLater(this);
}
```

File: src/main/java/rx/schedulers/JavaFxScheduler.java [hunk 3]
```diff
@@ -46,86 +48,167 @@ public static JavaFxScheduler getInstance() {
...
+                if (trampoline(runnable)) {
+                    return Subscriptions.unsubscribed();
                 }
+                else {
+                    queue.offer(runnable);
+                    EventQueue.invokeLater(this);
+                }
+            }
 
+            tracking.add(s);
```
Comment (akarnokd) [inline src/main/java/rx/schedulers/JavaFxScheduler.java]: This has to happen before the task is scheduled, otherwise you may add a dead task on arrival.

**Comment by akarnokd** [comment]
Looks like you didn't properly match the original Swing scheduler, see my comments inline.

**Comment by thomasnield** [comment]
Sorry, it looks like my GIT push failed last night when I walked away from my computer. Latest version is up. I really appreciate your help @akarnokd

**Comment by akarnokd** [comment]
There is still one comment unaddressed.

**Comment by thomasnield** [comment]
Ah, I completely misplaced that `else` block. I put it on the inner `if`. 

Instead of this:

``` java
if (Platform.isFxApplicationThread()) {
    if (trampoline(runnable)) {
        return Subscriptions.unsubscribed();
    }
    else {
        queue.offer(runnable);
        Platform.runLater(this);
    }
}
```

it should be this:

``` java
if (Platform.isFxApplicationThread()) {
    if (trampoline(runnable)) {
        return Subscriptions.unsubscribed();
    }
} else {
    queue.offer(runnable);
    Platform.runLater(this);
}
```

It seems to work now! Tests have passed and the `observeOn()` is now working. I'm going to do one more sanity check and then I'm going to merge. Thanks for your time and help @akarnokd!

**Comment by thomasnield** [comment]
This trampoline mechanism is brilliant. I didn't understand it too well until now. Now no matter how I try to break it, everything still works perfectly. Thanks again @akarnokd 

``` java
@Test
public void testRxObservableListAdds() {
    new JFXPanel();

    ObservableList<String> sourceList = FXCollections.observableArrayList();
    Observable<String> emissions = JavaFxObservable.fromObservableListAdds(sourceList);

    Platform.runLater(() -> {

        Observable<Integer> count = emissions.observeOn(Schedulers.io())
                .observeOn(JavaFxScheduler.getInstance())
                .take(3)
                .count();

        sourceList.add("Alpha");
        sourceList.add("Beta");
        sourceList.add("Gamma");

        assertTrue(count.toBlocking().first() == 3);  //successful!

    });
}
``
```

**Pull request merged**